### PR TITLE
1409: PSSS Confirmed Weekly Report

### PR DESCRIPTION
**Pull request by rohan-bes**
### Issue https://github.com/beyondessential/tupaia-backlog/issues/1409:

Added `Confirmed Weekly Report` Report

### CODE REVIEW

File: packages/report-server/src/types.ts [hunk 1]
```diff
@@ -18,9 +18,19 @@ export interface FetchReportParams extends ParamsDictionary {
...
+    dataGroups?: string[];
+  };
+  transform: (string | Record<string, unknown>)[];
+}
+
 interface ReportsRequestBody {
   emailAddress: string;
   password: string;
+  testConfig?: ReportConfig;
+  testData?: Record<string, string | number>[];
```
Comment (rohan-bes) [inline packages/report-server/src/types.ts:33]: Added these two optional parameters, which allow to provide either a `testConfig` object (representing a report.config object) or `testData` representing an array or analytics or events. These can be really useful for testing when data isn't available, or for iteratively developing a transform config without needing to edit the database.

I'm thinking that ultimately it might be nice to break this out into a separate endpoint (perhaps `/report/:id/fetch` and `/report/:id/test`) but I still haven't made my mind up about it yet

File: packages/report-server/src/routes/fetchReport.ts [hunk 2]
```diff
@@ -18,7 +18,19 @@ class FetchReportRouteHandler {
...
+      const reportBuilder = new ReportBuilder(
+        body.testConfig,
+        this.aggregator,
+        query,
+        body.testData,
+      );
+      const data = await reportBuilder.build();
+      respond(res, data, 200);
+      return;
+    }
```
Comment (kael89) [inline packages/report-server/src/routes/fetchReport.ts:32]: Let's encapsulate this in a separate function

File: packages/report-server/src/types.ts (hunk 1 above)
Comment (kael89) [reply packages/report-server/src/types.ts:33]: A separate endpoint would be great for keeping things clean and decoupled! As long as the actual behaviour can easily be invoked with test data passed in (to avoid duplication)

I guess your current implementation is still pretty clean, so we may as well leave as is for now

File: packages/report-server/src/types.ts [hunk 3]
```diff
@@ -40,7 +50,5 @@ export interface Event {
   eventDate: string;
   orgUnitName: string;
   orgUnit: string;
-  dataValues?: {
-    [key: string]: string | number;
-  };
+  dataValues?: Record<string, string | number>;
```
Comment (kael89) [inline packages/report-server/src/types.ts:53]: 👍

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+var dbm;
+var type;
+var seed;
+
+const permissionGroupNameToId = async (db, name) => {
+  const record = await db.runSql(`SELECT id FROM permission_group WHERE name = '${name}'`);
+  return record.rows[0] && record.rows[0].id;
+};
+
+const REPORT_CODE = 'PSSS_CONFIRMED_WEEKLY_REPORT';
```
Comment (kael89) [inline packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:14]: Perhaps we want to use "first letter upper" snake case for report codes? `PSSS_Confirmed_Weekly_Report`, similar to existing `dashboardReport` ids

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+          'PSSS_Confirmed_PF_Alert_Threshold_Crossed',
+          'PSSS_Confirmed_DLI_Cases',
+          'PSSS_Confirmed_DLI_Alert_Threshold_Crossed',
+        ],
+      },
+      transform: [
+        'keyValueByDataElementName',
+        {
+          transform: 'select',
+          "'period'": "convertToPeriod($row.period, 'WEEK')",
```
Comment (kael89) [inline packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:51]: Does this need to be wrapped in single quotes? Or would the following also work?

```suggestion
          "period": "convertToPeriod($row.period, 'WEEK')",
```

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+          '...': '*',
```
Comment (kael89) [inline packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:52]: I can see this happening a lot, so we could consider making it the default behaviour (no harm keeping additional fields, even if they are eventually not needed)

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        {
+          transform: 'aggregate',
+          period: 'group',
+          organisationUnit: 'group',
+          '...': 'last',
+        },
+        {
+          transform: 'select',
+          "'AFR'": '$row.PSSS_Confirmed_AFR_Cases',
+          "'AFR Threshold Crossed'": '$row.PSSS_Confirmed_AFR_Alert_Threshold_Crossed == 1',
```
Comment (kael89) [inline packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:63]: I'm not sure if our front-end expects a boolean value here, or `1`/`0`, or `Yes`/`No`

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+          'PSSS_Confirmed_ILI_Cases',
+          'PSSS_Confirmed_ILI_Alert_Threshold_Crossed',
+          'PSSS_Confirmed_PF_Cases',
+          'PSSS_Confirmed_PF_Alert_Threshold_Crossed',
+          'PSSS_Confirmed_DLI_Cases',
+          'PSSS_Confirmed_DLI_Alert_Threshold_Crossed',
+        ],
+      },
+      transform: [
+        'keyValueByDataElementName',
```
Comment (kael89) [inline packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:48]: I love this succinct transformation!

**Review by kael89 — CHANGES_REQUESTED** [review]
Excellent work!! I love the new report config, it's very declarative!

So good to see all this work coming together!! 👏🚀

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js (hunk 5 above)
Comment (rohan-bes) [reply packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:51]: So, the wrapping of keys in the `select` transform in quotes is because these keys are actually run through the `expression-parser` in order to evaluate their value. Doing this allows for some useful tricks like being able to key one value in a row by another value, eg. `"$row.dataElement" : "$row.value"`

However, this is obviously not the common case, so really we should have some logic in place to know whether or not we want to run the key through the `expression-parser`. I just haven't implemented that as of yet 😝 

I would like it do it soon tho, and if you don't wanna create any reports until we've got this done I can understand.

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js (hunk 7 above)
Comment (rohan-bes) [reply packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:63]: As discussed since this is for `PSSS` we can pass through a boolean value 👍

File: packages/report-server/src/types.ts (hunk 1 above)
Comment (rohan-bes) [reply packages/report-server/src/types.ts:33]: Yeah I think both endpoints would use the same `ReportBuilder` class, just only one has the ability to add the test config / test data. 

The idea around splitting into separate endpoints is partially cos the decoupling would be nice, but also a security thing. Regular users should be able to fetch data, but only Admin users (or some other group) should be able to test different configs out

File: packages/report-server/src/reportBuilder/reportBuilder.ts [hunk 9]
```diff
@@ -9,22 +9,33 @@ interface BuildReport {
...
 
-  constructor(report, aggregator: Aggregator, query: FetchReportQuery) {
-    this.report = report;
+  readonly testData: Row[];
+
+  constructor(
+    config: ReportConfig,
+    aggregator: Aggregator,
+    query: FetchReportQuery,
+    testData: Row[] = [],
```
[the remark this answers is not in the document]
Comment (rohan-bes) [reply packages/report-server/src/reportBuilder/reportBuilder.ts]: Yeah I liked this idea, so I gave it a shot in the latest commit.

File: packages/report-server/src/types.ts (hunk 1 above)
Comment (kael89) [reply packages/report-server/src/types.ts:33]: Most definitely!

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        'convertPeriodToWeek',
+        'aggregateFirstValuePerPeriodPerOrgUnit',
```
Comment (rohan-bes) [inline packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:50]: I went with this `aggregateFirstValuePerPeriodPerOrgUnit` as there really only should be 1 value per dataElement per week. This is part of the business logic of PSSS. It's possible to use a `mostRecent` type aggregation here but those really ought to be proceeded with a `sort` which is a bit tricky to do here as we want to transform the period type, and also is a little inefficient.

Given that it's a case that's not meant to happen, I'm happy with going with `first` here (a bit like when you know there should just be a single matching value in an array you use `array.find()`) 👍

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js (hunk 5 above)
Comment (kael89) [reply packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js]: No worries, it's not big deal! It's great that we have a plan for this

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js (hunk 10 above)
Comment (kael89) [reply packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:50]: Sounds good to me! How about naming it `keepFirstValuePerPeriodPerOrgUnit` or `pickFirstValuePerPeriodPerOrgUnit`

File: packages/report-server/src/reportBuilder/reportBuilder.ts [hunk 11]
```diff
@@ -9,33 +9,25 @@ interface BuildReport {
...
+
+  setTestData = (testData: Row[]) => {
     this.testData = testData;
-  }
+  };
 
-  build = async (): Promise<BuildReport> => {
+  build = async (aggregator: Aggregator, query: FetchReportQuery): Promise<BuildReport> => {
+    if (!this.config) {
+      throw new Error('Report requires a config be set');
```
Comment (kael89) [inline packages/report-server/src/reportBuilder/reportBuilder.ts:26]: Nice!

File: packages/report-server/src/reportBuilder/transform/aliases/aggregateAliases.ts [hunk 12]
```diff
@@ -12,3 +12,13 @@ export const aggregateMostRecentValuePerOrgUnit = () =>
...
+
+export const aggregateFirstValuePerPeriodPerOrgUnit = () =>
+  buildTransform([
+    {
+      transform: 'aggregate',
+      organisationUnit: 'group',
+      period: 'group',
+      '...': 'first',
+    },
+  ]);
```
Comment (kael89) [inline packages/report-server/src/reportBuilder/transform/aliases/aggregateAliases.ts:24]: So easy to add! That's a sign of a great architecture under the hood 😉

File: packages/report-server/src/routes/fetchReport.ts [hunk 13]
```diff
@@ -19,51 +19,66 @@ class FetchReportRouteHandler {
...
+  models,
+  accessPolicy,
+  report,
+  organisationUnitCodes: string[],
+) => {
+  const permissionGroup = await models.permissionGroup.findById(report.permission_group_id);
+  const orgUnitsAndCodes = await Promise.all(
+    organisationUnitCodes.map(async orgUnitCode => ({
+      orgUnitCode,
+      orgUnit: await models.entity.findOne({ code: orgUnitCode }),
```
Comment (kael89) [inline packages/report-server/src/routes/fetchReport.ts:64]: Could we change this to be a single query, as discussed in the past? (Can do in a separate PR)

File: packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregations.ts [hunk 14]
```diff
@@ -54,6 +54,12 @@ const drop = (existingRow: Row, field: string, value: FieldValue) => {
   // Do nothing, don't add the field to the existing row
 };
 
+const first = (existingRow: Row, field: string, value: FieldValue) => {
+  if (!existingRow[field]) {
+    existingRow[field] = value;
+  }
```
Comment (kael89) [inline packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregations.ts:60]: (value may be `0`)

```suggestion
  if (!(field in existingRow)) {
    existingRow[field] = value;
  }
```

**Review by kael89 — CHANGES_REQUESTED** [review]
I love the new report config, so simple! Almost there

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js (hunk 10 above)
Comment (rohan-bes) [reply packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:50]: I was prefixing the names of all these aliases for `aggregate` transformations with `aggregate` as a bit of a hint to the user that 'hey, this is done using aggregation just so ya know` 😛 

But I do agree it's a bit awkward 👍 

How would you feel about just simply: `firstValuePerPeriodPerOrgUnit`? Then we could also have `sumValuesPerPeriod`, or `minValuePerOrgUnit`, etc. ? Or should it really have a verb in front for it to read like a transformation..?

File: packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregations.ts (hunk 14 above)
Comment (rohan-bes) [reply packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregations.ts:60]: 🤦 why do I still always do this!! Thanks, great catch!!

Comment (kael89) [reply packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregations.ts:60]: I've been burned by that countless times...

File: packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js (hunk 10 above)
Comment (kael89) [reply packages/database/src/migrations/20201123061122-AddConfirmedWeeklyDataReportForPSSS-modifies-data.js:50]: I really like `firstValuePerPeriodPerOrgUnit`, reads very natural!

File: packages/report-server/src/routes/fetchReport.ts (hunk 13 above)
Comment (rohan-bes) [reply packages/report-server/src/routes/fetchReport.ts]: Yup, good point, did this 👍

File: packages/report-server/src/routes/fetchReport.ts [hunk 15]
```diff
@@ -55,25 +55,21 @@ const checkUserHasAccessToReport = async (
...
+  const foundOrgUnits = await models.entity.find({ code: requestedOrgUnitCodes });
+  const foundOrgUnitCodes = foundOrgUnits.map(orgUnit => orgUnit.code);
 
-  const countryCodes = new Set(
-    orgUnitsAndCodes.map(orgUnitAndCode => orgUnitAndCode.orgUnit.country_code),
+  const missingOrgUnitCodes = requestedOrgUnitCodes.filter(
+    orgUnitCode => !foundOrgUnitCodes.includes(orgUnitCode),
   );
+  if (missingOrgUnitCodes.length > 0) {
+    throw new Error(`No entities found with codes ${missingOrgUnitCodes}`);
```
Comment (kael89) [inline packages/report-server/src/routes/fetchReport.ts:69]: Great implementation!

**Review by kael89 — APPROVED** [review]
Looks great! A historical moment, the first revamped `report` is ready to be merged! 🚀
